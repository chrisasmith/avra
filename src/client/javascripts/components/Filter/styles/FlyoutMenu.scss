@import '../../../../stylesheets/modules/theme.scss';

.flyout {
  position: absolute;
  top: 50px;
  left: 216px;
  background-color:#fff;
  height: 700px;
  width: 290px;
  z-index: 10000;
  overflow: hidden;
  border-top: 1px solid #ccc;
  border-right: 1px solid #ccc;
  border-bottom: 1px solid #ccc;
  animation: fadeIn .25s ease-out;
}

.page {
  width: 100%;
  height: 100%;
  position: absolute;
  overflow-y: scroll;
  padding-bottom: 16px;
  padding-top: 16px;
}

.pageEnter {
  transform: translateX(100%);
}

.pageEnterActive {
  transform: translateX(0%);
  transition: all 300ms ease-out;
}

.pageExit {
  transform: translateX(0);
}

.pageExitActive {
  transform: translateX(-100%);
  transition: all 300ms ease-out;
}

.top {
  border-bottom: 1px solid #ccc;
  padding-top: .5rem;
  margin: 0 24px .5rem 24px;
}

.item {
  padding-left: 34px;
  padding-right: 16px;
  color: $dark-gray-color;

  &:hover {
    cursor: pointer;
    background-color: #ccc;
    color: #fff;
  }
}

.next {
  float: right;
  font-size: 12px;
  line-height: 22px;
}

@keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
