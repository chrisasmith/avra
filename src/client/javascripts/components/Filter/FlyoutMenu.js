import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { CSSTransition } from 'react-transition-group'
import styles from './styles/FlyoutMenu.scss'

const propTypes = {
  options: PropTypes.array.isRequired,
  onItemSelected: PropTypes.func.isRequired,
}

class FlyoutMenu extends Component {
  constructor(props, context) {
    super(props, context)

    this.onSelectItem = this.onSelectItem.bind(this)
    this.onGlobalClick = this.onGlobalClick.bind(this)
    this.self = React.createRef()

    this.state = {
      page: 0,
      items: [
        [...props.options],
        [],
        [],
      ],
    }
  }

  componentDidMount() {
    document.addEventListener('click', this.onGlobalClick, false)
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.onGlobalClick, false)
  }

  onGlobalClick(e) {
    // bail if this is a detached element
    // this can happen if we removed the element we clicked on, generally
    // that means we didn't click outside the popup
    if (!document.body.contains(e.target)) {
      return
    }

    // check to see if we contain this click
    const domElement = this.self.current
    if (domElement && !domElement.contains(e.target)) {
      this.props.onItemSelected()
    }
  }


  onSelectItem(item) {
    if (item.children) {
      const page = this.state.page + 1
      const items = [...this.state.items]
      items[page] = [...item.children]

      this.setState({ items, page })
    } else {
      this.props.onItemSelected(item)
    }
  }

  render() {
    return (
      <div className={styles.flyout} ref={this.self}>
        {[0, 1, 2].map(page => (
          <CSSTransition
            key={page}
            in={this.state.page === page}
            timeout={300}
            classNames={{
              enter: styles.pageEnter,
              enterActive: styles.pageEnterActive,
              exit: styles.pageExit,
              exitActive: styles.pageExitActive,
            }}
            unmountOnExit
          >
            <div onClick={this.onPageClick} className={styles.page}>
              {this.state.items[page].map(item => (
                <div key={item.id}>
                  <div
                    onClick={() => this.onSelectItem(item)}
                    className={classnames('pt-2 pb-2', styles.item)}
                  >
                    {item.name}
                    {item.children && (
                      <i className={classnames('fa fa-chevron-right', styles.next)} />
                    )}
                  </div>
                  <div className={classnames({ [styles.top]: item.isGroup })} />
                </div>
              ))}
            </div>
          </CSSTransition>
        ))}
      </div>
    )
  }
}

FlyoutMenu.propTypes = propTypes
export default FlyoutMenu
