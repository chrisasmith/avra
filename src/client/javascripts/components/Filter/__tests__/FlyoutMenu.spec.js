import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import FlyoutMenu from '../FlyoutMenu'

describe('Components FlyoutMenu', () => {
  it('should render', () => {
    const wrapper = shallow(<FlyoutMenu
      options={[]}
      onItemSelected={jest.fn()}
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should render items', () => {
    const wrapper = shallow(<FlyoutMenu
      options={[{ name: 'hi', id: 0 }]}
      onItemSelected={jest.fn()}
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should select an item', () => {
    const onItemSelected = jest.fn()
    const item = { name: 'hi', id: 0 }
    const wrapper = shallow(<FlyoutMenu
      options={[item]}
      onItemSelected={onItemSelected}
    />)

    wrapper.find('.item').simulate('click')
    expect(onItemSelected).toHaveBeenCalledWith(item)
  })

  it('should render group items', () => {
    const wrapper = shallow(<FlyoutMenu
      options={[{ name: 'hi', id: 0, isGroup: true }]}
      onItemSelected={jest.fn()}
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should render items with children', () => {
    const wrapper = shallow(<FlyoutMenu
      options={[{ name: 'hi', id: 0, children: [{ name: 'second', id: 1 }] }]}
      onItemSelected={jest.fn()}
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should switch pages', () => {
    const wrapper = shallow(<FlyoutMenu
      options={[{ name: 'hi', id: 0, children: [{ name: 'second', id: 1 }] }]}
      onItemSelected={jest.fn()}
    />)

    wrapper.find('.item').simulate('click')
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
})
