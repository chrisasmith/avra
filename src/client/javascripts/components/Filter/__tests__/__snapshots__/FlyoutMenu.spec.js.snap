// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components FlyoutMenu should render 1`] = `
<div
  className="flyout"
>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={true}
    key="0"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    />
  </CSSTransition>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={false}
    key="1"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    />
  </CSSTransition>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={false}
    key="2"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    />
  </CSSTransition>
</div>
`;

exports[`Components FlyoutMenu should render group items 1`] = `
<div
  className="flyout"
>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={true}
    key="0"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    >
      <div
        key="0"
      >
        <div
          className="pt-2 pb-2 item"
          onClick={[Function]}
        >
          hi
        </div>
        <div
          className="top"
        />
      </div>
    </div>
  </CSSTransition>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={false}
    key="1"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    />
  </CSSTransition>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={false}
    key="2"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    />
  </CSSTransition>
</div>
`;

exports[`Components FlyoutMenu should render items 1`] = `
<div
  className="flyout"
>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={true}
    key="0"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    >
      <div
        key="0"
      >
        <div
          className="pt-2 pb-2 item"
          onClick={[Function]}
        >
          hi
        </div>
        <div
          className=""
        />
      </div>
    </div>
  </CSSTransition>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={false}
    key="1"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    />
  </CSSTransition>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={false}
    key="2"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    />
  </CSSTransition>
</div>
`;

exports[`Components FlyoutMenu should render items with children 1`] = `
<div
  className="flyout"
>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={true}
    key="0"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    >
      <div
        key="0"
      >
        <div
          className="pt-2 pb-2 item"
          onClick={[Function]}
        >
          hi
          <i
            className="fa fa-chevron-right next"
          />
        </div>
        <div
          className=""
        />
      </div>
    </div>
  </CSSTransition>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={false}
    key="1"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    />
  </CSSTransition>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={false}
    key="2"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    />
  </CSSTransition>
</div>
`;

exports[`Components FlyoutMenu should switch pages 1`] = `
<div
  className="flyout"
>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={false}
    key="0"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    >
      <div
        key="0"
      >
        <div
          className="pt-2 pb-2 item"
          onClick={[Function]}
        >
          hi
          <i
            className="fa fa-chevron-right next"
          />
        </div>
        <div
          className=""
        />
      </div>
    </div>
  </CSSTransition>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={true}
    key="1"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    >
      <div
        key="1"
      >
        <div
          className="pt-2 pb-2 item"
          onClick={[Function]}
        >
          second
        </div>
        <div
          className=""
        />
      </div>
    </div>
  </CSSTransition>
  <CSSTransition
    classNames={
      Object {
        "enter": "pageEnter",
        "enterActive": "pageEnterActive",
        "exit": "pageExit",
        "exitActive": "pageExitActive",
      }
    }
    in={false}
    key="2"
    timeout={300}
    unmountOnExit={true}
  >
    <div
      className="page"
    />
  </CSSTransition>
</div>
`;
